Api v3 groups remaining routes

TheHollidayInn opened the pull request:
Working on more group routes here. I also added some minor fixes for linting and for the test stops, but I can remove the test stop since it has been fixed. 

Also, I realized that I needed the invite and join routes for the next update test, so I marked it as pending and will pick up there tomorrow. Feel free to look over the code until then.

TheHollidayInn committed:
Added get and list group tests

TheHollidayInn committed:
Added expect to global eslint config

paglias commented on test/api/v3/integration/groups/PUT-groups.test.js:
```diff
@@ -0,0 +1,64 @@
+    });
+  });
+
+  xit('returns an error when a non group leader tries to update', () => {
```
@TheHollidayInn why is this test skipped?

TheHollidayInn replied:
I just stopped here for the day, and posted it in case you wanted to review. I'm about to reset and force push to overwrite the commit. 

The reason I stopped is because I realized I needed a second member in the group to try and make an edit. So, I am assuming I need to call the invite and join routes for the second user. But, from Slack, it seems there may be a better way. Do you have any suggestions?

TheHollidayInn committed:
Added initial group update tests

crookedneighbor commented on .eslintrc:
```diff
@@ -113,6 +113,9 @@
         "mocha": true,
         "node": true
     },
+    "globals": {
```
This is not necessary. 

It's checked here: https://github.com/HabitRPG/habitrpg/blob/5304b37102f7210b00f63fdefe4041f8ec08a720/tasks/gulp-eslint.js#L61-L65

Going to move these rules to their own `.eslintrc` in the tests directory soon so a normal `eslint path/to/file` will work correctly.

TheHollidayInn replied:
Sounds good. I was having this issue when running eslint at the time.

crookedneighbor replied:
Please remove these.

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,129 @@
+  let user;
+  let api;
+
+  beforeEach(() => {
```
This should be updated to use `async` and `await`

TheHollidayInn replied:
Yea, I had these posted before our discussion. I will be updating them today.

crookedneighbor commented on test/api/v3/integration/groups/PUT-groups.test.js:
```diff
@@ -0,0 +1,57 @@
+    let groupUpdatedName = 'Test Public Guild Updated';
+    let newUser = await generateUser();
+
+    return user.post(`/groups/${groupToUpdate._id}/invite`, {
```
All of these should be updated to use await instead of promise chains.

TheHollidayInn committed:
Updated get group tests to use await/async syntax

TheHollidayInn commented:
Okay, I made some changes. I still have the PUT reject test pending until I push my invite route. I am almost done with the invite route.

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,101 @@
+  translate as t,
+} from '../../../../helpers/api-integration.helper';
+
+describe('GET /group', () => {
```
This should be `Get /groups` (plural)

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,101 @@
+  });
+
+  it('returns an error when group is not found', async () => {
+    await expect(user.get('/groups/fakeGroupId', {}))
```
This file is for testing `GET /groups` and this test is testing `GET /groups/:id`

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,101 @@
+          name: groupName,
+          type: groupType,
+        });
+        let groupFound = await user.get(`/groups/${createdGroup._id}`);
```
All the rest of the tests in this file are testing `GET /groups/:id`. They should be moved to their own test file.

TheHollidayInn committed:
Converted POST and PUT routes to await/async syntax

TheHollidayInn committed:
Separated /groups/:id route to its own test file

TheHollidayInn commented:
Alright, I fixed lint errors from the previous tests and moved the group id tests. Then, I added my progress on invites which I will be overwriting with force push soon. But, if you want to take a look at what I am doing so far, feel free.

TheHollidayInn committed:
Added invite route and initial tests

KristianTashkov commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -391,145 +392,101 @@ api.removeGroupMember = {
+    if (_.contains(userToInvite.guilds, group._id)) {
+      throw new NotAuthorized(res.t('userAlreadyInGroup'));
     }
+    if (userToInvite.invitations && userToInvite.invitations.guilds && _.find(userToInvite.invitations.guilds, {id: group._id})) {
```
Aren't `userToInvite.invitations` and `userToInvite.invitations.guilds` part of the model schema? Can they ever be undefined?

paglias replied:
good point, they'll always be defined

paglias commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -563,15 +520,25 @@ api.inviteToGroup = {
+      results.push(emailResults);
+    }
+
+    res.respond(200, {result: results});
```
@TheHollidayInn what are we sending to the client here exactly?

TheHollidayInn replied:
These just return the values that wen through the Qs up top. Basically, these just return the return values from _inviteByUUID and _inviteByEmail which are just the information from uuids, emails. So, it is probably useless. I was using the info to confirm that the information passed.

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -4,14 +4,15 @@ import _ from 'lodash';
 import * as firebase from '../../libs/api-v3/firebase';
-import { txnEmail } from '../../libs/api-v3/email';
-// import { encrypt } from '../../libs/api-v3/encryption';
+import { sendTxn } from '../../libs/api-v3/email';
```
Let's  make this `import { sendTxn as sendTxnEmail }` so it's clear what this function is doing when we use it.

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,51 @@
+      type: groups[2].type,
+    });
+
+    expect(group1).toExist;
```
These three expectations shouldn't be necessary. If there's an error, it'll get thrown and mocha will tell you about it. If it's not actually making  the groups, that's a concern of the post route, not this test.

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,51 @@
+      },
+    ];
+
+    let group1 = await user.post('/groups', {
```
We don't actually care about retaining a reference the the groups here, so why not:

```
groups.forEach(async (groupData) => {
  await user.post('/groups', groupData);
});
```

crookedneighbor replied:
Updated to actually work.

paglias replied:
forEach won't block the main function execution.

Better do `await Q.all(groups.map(group => await user.post(...)))`

crookedneighbor replied:
Good call.

TheHollidayInn replied:
Yea, I actually had forEach, but then that was an eslint error. I didn't know how to use Q.all at the time, but I will make the update.

paglias replied:
@TheHollidayInn forEach did produce an error because when you use await the direct parent function must be `async function () {}` not a normal `function () {}`

TheHollidayInn replied:
Hmm. I don't exactly remember my code, but the issue was the eslint said not to use await inside of a for loop. I'll take a screenshot if it comes up again, though.

crookedneighbor commented:
In my opinion, these should be separate PRs

TheHollidayInn commented:
Yea, I can see how this should be separate PRs. The issue for me is that a lot of the routes and tests depend on each other, so it is hard to start working on the next route without the previous code.

TheHollidayInn commented:
Okay, I made the updates.

TheHollidayInn committed:
Updated test to use Q, removed eslint config, and other minor edits

TheHollidayInn committed:
Added initial leave tests and updated group leave function

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,35 @@
+import {
```
Shouldn't this test look more like: https://github.com/HabitRPG/habitrpg/blob/8a546c7ec3d41acd606991aca1f78c813f0ff64b/test/api/v2/groups/GET-groups.test.js

Adjusted of course for the way v3 differs from v2.

crookedneighbor commented on test/api/v3/integration/groups/GET-groups_id.test.js:
```diff
@@ -0,0 +1,73 @@
+import {
```
See https://github.com/HabitRPG/habitrpg/blob/8a546c7ec3d41acd606991aca1f78c813f0ff64b/test/api/v2/groups/GET-groups_id.test.js

crookedneighbor commented on test/api/v3/integration/groups/GET-groups.test.js:
```diff
@@ -0,0 +1,35 @@
+import {
+  generateUser,
+} from '../../../../helpers/api-integration.helper';
```
This will need to change to ../../../../helpers/api-v3-integration.helper';

crookedneighbor commented on test/api/v3/integration/groups/GET-groups_id.test.js:
```diff
@@ -0,0 +1,73 @@
+import {
+  generateUser,
+  translate as t,
+} from '../../../../helpers/api-integration.helper';
```
This will need to change to ../../../../helpers/api-v3-integration.helper';

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_leave.js:
```diff
@@ -0,0 +1,105 @@
+import {
+  generateUser,
+} from '../../../../helpers/api-integration.helper';
```
This will need to change to ../../../../helpers/api-v3-integration.helper';

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -0,0 +1,197 @@
+import {
+  generateUser,
+  translate as t,
+} from '../../../../helpers/api-integration.helper';
```
This will need to change to ../../../../helpers/api-v3-integration.helper';

Also, we need a line break.

crookedneighbor commented on test/api/v3/integration/groups/PUT-groups.test.js:
```diff
@@ -0,0 +1,55 @@
+import {
+  generateUser,
+  translate as t,
+} from '../../../../helpers/api-integration.helper';
```
This will need to change to ../../../../helpers/api-v3-integration.helper';
